### PR TITLE
nixos/bepasty: Kind of modernize the module

### PR DESCRIPTION
**Pull request by dasJ**
- Switch to Python 3 which fixes bepasty not being found (and requires a
patch to work which is already on bepasty master)
- Remove the enable option
- Hardcode directories and let systemd manage permissions
- Sandbox the service using systemd
- Switch to a dynamic user and run nothing as root (except for preStart)
- Allow users to set permissions without adding them to the store
(extraConfigFile)
- Make assertion more clear

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Lassulus** [comment]
@GrahamcOfBorg build bepasty

File: pkgs/tools/misc/bepasty/default.nix [hunk 1]
```diff
@@ -11,6 +11,15 @@ buildPythonPackage rec {
   pname = "bepasty";
   version = "0.5.0";
 
+  patches = [
+    # Fix for bytes under Python 3
+    # See https://github.com/bepasty/bepasty-server/issues/200
+    (fetchpatch {
```
Comment (Lassulus) [inline pkgs/tools/misc/bepasty/default.nix]: fetchpatch is missing

Comment (dasJ) [reply pkgs/tools/misc/bepasty/default.nix]: Fixed

**Comment by Lassulus** [comment]
@GrahamcOfBorg build bepasty

File: nixos/modules/services/misc/bepasty.nix [hunk 2]
```diff
@@ -118,10 +105,15 @@ in
     };
   };
 
-  config = mkIf cfg.enable {
+  config = mkIf (length (attrNames cfg.servers) > 0) {
```
Comment (Lassulus) [inline nixos/modules/services/misc/bepasty.nix]: how about: `cfg.server != {}` ?

Comment (dasJ) [reply nixos/modules/services/misc/bepasty.nix]: Great idea, implemented.

**Comment by aanderse** [comment]
@dasJ After a quick review it looks like this will make existing installs lose all data if custom directories were specified. Is this correct?

File: nixos/modules/services/misc/bepasty.nix [hunk 3]
```diff
@@ -136,48 +128,50 @@ in
               extraLibs = [ bepasty gevent ];
             };
           in {
-            BEPASTY_CONFIG = "${server.workDir}/bepasty-${name}.conf";
+            BEPASTY_CONFIG = "/tmp/bepasty.conf";
```
Comment (memberbetty) [inline nixos/modules/services/misc/bepasty.nix]: You cannot assume that /tmp/bepasty.conf doesn't exist yet.

Comment (memberbetty) [reply nixos/modules/services/misc/bepasty.nix]: You seem to have removed `PrivateTmp= enabled`. If you add that again, it might work as you intended.

Comment (dasJ) [reply nixos/modules/services/misc/bepasty.nix]: No worries, `DynamicUser` also enables `PrivateTmp`

**Comment by dasJ** [comment]
@aanderse Yes, it's a a little downside. On the upside, systemd will manage the permissions of the directory and make the path writable in the sandbox.

**Comment by aanderse** [comment]
@dasJ You might want to consider keeping the `dataDir` and `workDir` option around but adding a warning that the options are deprecated. This way existing installs could be automatically migrated. In a few NixOS releases the options could eventually be removed. If you do this then no one will lose any data, or have to do any migration work.

**Comment by dasJ** [comment]
@aanderse The problem is: How do I ensure the directories are owned by the proper user and have the proper permissions?

**Comment by aanderse** [comment]
I assume some scripts could be written to take the existing `dataDir` and `workDir` values which aren't null and move the data to the appropriate spot, but I haven't looked into it in depth.

**Pull request closed without merging**